File: src/cache/marketplace.rs
//! Normalize a single-plugin marketplace repository into the package-root layout used by the
//! resolver and staging pipeline.
//!
//! Some publishers keep one logical plugin in per-harness directories and expose those directories
//! through marketplace manifests at the repository root. Agentpack locks the repository root as one
//! package, so this module resolves the local marketplace sources once and merges them into a
//! canonical cache root before ordinary plugin classification runs.

use std::collections::{BTreeMap, BTreeSet, HashSet};
use std::fs;
use std::path::{Component, Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};

use serde::Deserialize;
use serde_json::Value;

use crate::error::{AgentpackError, Result};
use crate::fs_util::{copy_merge_tree, read_json_value};

const CLAUDE_MARKETPLACE: &str = ".claude-plugin/marketplace.json";
const CURSOR_MARKETPLACE: &str = ".cursor-plugin/marketplace.json";
const CODEX_MARKETPLACE: &str = ".agents/plugins/marketplace.json";

pub(super) const MARKETPLACE_MANIFESTS: [&str; 3] =
    [CLAUDE_MARKETPLACE, CURSOR_MARKETPLACE, CODEX_MARKETPLACE];

const PLUGIN_MANIFESTS: [&str; 3] = [
    ".claude-plugin/plugin.json",
    ".cursor-plugin/plugin.json",
    ".codex-plugin/plugin.json",
];

#[derive(Clone, Copy)]
struct MarketplaceSpec {
    relative_path: &'static str,
    // Sources are merged from least to most preferred. Claude is the portable baseline agentpack
    // historically consumed; Cursor and Codex sources supplement it with target-specific files.
    priority: u8,
}

const MARKETPLACE_SPECS: [MarketplaceSpec; 3] = [
    MarketplaceSpec {
        relative_path: CODEX_MARKETPLACE,
        priority: 0,
    },
    MarketplaceSpec {
        relative_path: CURSOR_MARKETPLACE,
        priority: 1,
    },
    MarketplaceSpec {
        relative_path: CLAUDE_MARKETPLACE,
        priority: 2,
    },
];

#[derive(Default, Deserialize)]
struct MarketplaceDocument {
    #[serde(default)]
    metadata: MarketplaceMetadata,
    #[serde(default)]
    plugins: Vec<MarketplacePlugin>,
}

#[derive(Default, Deserialize)]
struct MarketplaceMetadata {
    #[serde(default, rename = "pluginRoot")]
    plugin_root: Option<String>,
}

#[derive(Deserialize)]
struct MarketplacePlugin {
    name: String,
    source: Value,
}

struct LocalPluginSource {
    path: PathBuf,
    priority: u8,
}

pub(super) fn cache_has_marketplace_manifest(root: &Path) -> bool {
    MARKETPLACE_MANIFESTS
        .iter()
        .any(|relative| root.join(relative).is_file())
}

pub(super) fn repo_dir_has_marketplace_manifest(paths: &HashSet<String>, dir: &str) -> bool {
    let dir = dir.trim_matches('/');
    MARKETPLACE_MANIFESTS.iter().any(|relative| {
        let path = if dir.is_empty() {
            (*relative).to_string()
        } else {
            format!("{dir}/{relative}")
        };
        paths.contains(&path)
    })
}

/// Materialize a marketplace root when it represents exactly one logical plugin.
///
/// Returns `false` when no marketplace exists. A marketplace containing multiple plugin names is
/// intentionally rejected: one `LockPackage` cannot truthfully represent an arbitrary catalog, and
/// silently flattening such a catalog would make artifact collisions order-dependent.
pub(super) fn materialize_single_plugin_marketplace(root: &Path) -> Result<bool> {
    if !cache_has_marketplace_manifest(root) {
        return Ok(false);
    }

    let canonical_root = fs::canonicalize(root).map_err(|err| AgentpackError::io(root, err))?;
    let mut plugin_names = BTreeSet::new();
    let mut local_sources: BTreeMap<PathBuf, u8> = BTreeMap::new();

    for spec in MARKETPLACE_SPECS {
        let manifest_path = root.join(spec.relative_path);
        if !manifest_path.is_file() {
            continue;
        }
        let document: MarketplaceDocument =
            serde_json::from_value(read_json_value(&manifest_path)?).map_err(|err| {
                AgentpackError::Cache(format!(
                    "invalid marketplace manifest {}: {err}",
                    manifest_path.display()
                ))
            })?;
        let mut names_in_document = HashSet::with_capacity(document.plugins.len());

        for plugin in document.plugins {
            let name = plugin.name.trim();
            if name.is_empty() {
                return Err(AgentpackError::Cache(format!(
                    "marketplace manifest {} contains a plugin with an empty name",
                    manifest_path.display()
                )));
            }
            if !names_in_document.insert(name.to_string()) {
                return Err(AgentpackError::Cache(format!(
                    "marketplace manifest {} contains duplicate plugin name {name:?}",
                    manifest_path.display()
                )));
            }
            plugin_names.insert(name.to_string());

            let Some(relative_source) = local_source_path(
                &plugin.source,
                document.metadata.plugin_root.as_deref(),
                &manifest_path,
                name,
            )?
            else {
                continue;
            };
            let source =
                canonical_local_source(&canonical_root, &relative_source, &manifest_path, name)?;
            validate_plugin_source(&source, name, &manifest_path)?;
            local_sources
                .entry(source)
                .and_modify(|priority| *priority = (*priority).max(spec.priority))
                .or_insert(spec.priority);
        }
    }

    if plugin_names.is_empty() {
        return Err(AgentpackError::Cache(format!(
            "marketplace at {} contains no plugins",
            root.display()
        )));
    }
    if plugin_names.len() != 1 {
        let names = plugin_names.into_iter().collect::<Vec<_>>().join(", ");
        return Err(AgentpackError::Cache(format!(
            "marketplace at {} contains multiple plugins ({names}); add a specific plugin source directory instead",
            root.display()
        )));
    }
    if local_sources.is_empty() {
        let name = plugin_names.into_iter().next().unwrap_or_default();
        return Err(AgentpackError::Cache(format!(
            "marketplace plugin {name:?} at {} has no local source; add its Git or package source directly",
            root.display()
        )));
    }

    let mut sources: Vec<LocalPluginSource> = local_sources
        .into_iter()
        .map(|(path, priority)| LocalPluginSource { path, priority })
        .collect();
    sources.sort_by(|a, b| {
        a.priority
            .cmp(&b.priority)
            .then_with(|| a.path.cmp(&b.path))
    });

    let normalized = unique_sibling(root, "marketplace-normalized")?;
    fs::create_dir_all(&normalized).map_err(|err| AgentpackError::io(&normalized, err))?;
    let merge_result = sources
        .iter()
        .try_for_each(|source| copy_merge_tree(&source.path, &normalized));
    if let Err(err) = merge_result {
        let _ = fs::remove_dir_all(&normalized);
        return Err(err);
    }

    if !PLUGIN_MANIFESTS
        .iter()
        .any(|relative| normalized.join(relative).is_file())
    {
        let _ = fs::remove_dir_all(&normalized);
        return Err(AgentpackError::Cache(format!(
            "marketplace at {} did not resolve to a plugin manifest",
            root.display()
        )));
    }

    replace_directory(root, &normalized)?;
    Ok(true)
}

fn local_source_path(
    source: &Value,
    plugin_root: Option<&str>,
    manifest_path: &Path,
    plugin_name: &str,
) -> Result<Option<PathBuf>> {
    let source_text = match source {
        Value::String(path) => Some(path.as_str()),
        Value::Object(object) => match object.get("source").and_then(Value::as_str) {
            Some("local") => Some(object.get("path").and_then(Value::as_str).ok_or_else(|| {
                AgentpackError::Cache(format!(
                    "local marketplace source for {plugin_name:?} in {} has no string path",
                    manifest_path.display()
                ))
            })?),
            _ => None,
        },
        _ => {
            return Err(AgentpackError::Cache(format!(
                "marketplace source for {plugin_name:?} in {} must be a string or object",
                manifest_path.display()
            )))
        }
    };
    let Some(source_text) = source_text else {
        return Ok(None);
    };

    // Official marketplace formats require `./` for an ordinary local source. Claude also permits
    // a short leaf such as `formatter` when `metadata.pluginRoot` supplies the base directory.
    if plugin_root.is_none() && !source_text.starts_with("./") {
        return Err(AgentpackError::Cache(format!(
            "local marketplace source {source_text:?} for {plugin_name:?} in {} must start with ./",
            manifest_path.display()
        )));
    }

    let mut relative = PathBuf::new();
    if let Some(base) = plugin_root {
        relative.push(safe_relative_path(base, manifest_path, plugin_name)?);
    }
    relative.push(safe_relative_path(source_text, manifest_path, plugin_name)?);
    if relative.as_os_str().is_empty() {
        return Err(AgentpackError::Cache(format!(
            "local marketplace source for {plugin_name:?} in {} resolves to the marketplace root",
            manifest_path.display()
        )));
    }
    Ok(Some(relative))
}

fn safe_relative_path(value: &str, manifest_path: &Path, plugin_name: &str) -> Result<PathBuf> {
    let path = Path::new(value);
    let mut clean = PathBuf::new();
    for component in path.components() {
        match component {
            Component::CurDir => {}
            Component::Normal(segment) => clean.push(segment),
            Component::ParentDir | Component::RootDir | Component::Prefix(_) => {
                return Err(AgentpackError::Cache(format!(
                    "unsafe marketplace source {value:?} for {plugin_name:?} in {}: paths must stay inside the marketplace root",
                    manifest_path.display()
                )))
            }
        }
    }
    Ok(clean)
}

fn canonical_local_source(
    canonical_root: &Path,
    relative_source: &Path,
    manifest_path: &Path,
    plugin_name: &str,
) -> Result<PathBuf> {
    let candidate = canonical_root.join(relative_source);
    let canonical = fs::canonicalize(&candidate).map_err(|err| {
        AgentpackError::Cache(format!(
            "cannot resolve marketplace source for {plugin_name:?} in {} at {}: {err}",
            manifest_path.display(),
            candidate.display()
        ))
    })?;
    if !canonical.starts_with(canonical_root) {
        return Err(AgentpackError::Cache(format!(
            "marketplace source for {plugin_name:?} in {} escapes the marketplace root: {}",
            manifest_path.display(),
            candidate.display()
        )));
    }
    if !canonical.is_dir() {
        return Err(AgentpackError::Cache(format!(
            "marketplace source for {plugin_name:?} in {} is not a directory: {}",
            manifest_path.display(),
            candidate.display()
        )));
    }
    Ok(canonical)
}

fn validate_plugin_source(source: &Path, expected_name: &str, marketplace: &Path) -> Result<()> {
    let mut found_manifest = false;
    for relative in PLUGIN_MANIFESTS {
        let manifest = source.join(relative);
        if !manifest.is_file() {
            continue;
        }
        found_manifest = true;
        let value = read_json_value(&manifest)?;
        let actual_name = value.get("name").and_then(Value::as_str).ok_or_else(|| {
            AgentpackError::Cache(format!(
                "plugin manifest {} referenced by {} has no string name",
                manifest.display(),
                marketplace.display()
            ))
        })?;
        if actual_name != expected_name {
            return Err(AgentpackError::Cache(format!(
                "marketplace plugin name {expected_name:?} in {} does not match manifest name {actual_name:?} at {}",
                marketplace.display(),
                manifest.display()
            )));
        }
    }
    if !found_manifest {
        return Err(AgentpackError::Cache(format!(
            "marketplace plugin {expected_name:?} in {} points to {}, which has no .claude-plugin, .cursor-plugin, or .codex-plugin manifest",
            marketplace.display(),
            source.display()
        )));
    }
    Ok(())
}

static UNIQUE_PATH_COUNTER: AtomicU64 = AtomicU64::new(0);

fn unique_sibling(path: &Path, label: &str) -> Result<PathBuf> {
    let parent = path.parent().ok_or_else(|| {
        AgentpackError::Cache(format!("cache root has no parent: {}", path.display()))
    })?;
    let name = path
        .file_name()
        .and_then(|value| value.to_str())
        .unwrap_or("package");
    loop {
        let serial = UNIQUE_PATH_COUNTER.fetch_add(1, Ordering::Relaxed);
        let candidate = parent.join(format!(
            ".agentpack-{label}-{name}-{}-{serial}",
            std::process::id()
        ));
        if !candidate.exists() {
            return Ok(candidate);
        }
    }
}

fn replace_directory(current: &Path, replacement: &Path) -> Result<()> {
    let backup = unique_sibling(current, "marketplace-backup")?;
    fs::rename(current, &backup).map_err(|err| AgentpackError::io(current, err))?;
    if let Err(err) = fs::rename(replacement, current) {
        let rollback = fs::rename(&backup, current);
        let _ = fs::remove_dir_all(replacement);
        if let Err(rollback_err) = rollback {
            return Err(AgentpackError::Cache(format!(
                "failed to install normalized marketplace cache ({err}) and rollback {} ({rollback_err})",
                current.display()
            )));
        }
        return Err(AgentpackError::io(current, err));
    }
    fs::remove_dir_all(&backup).map_err(|err| AgentpackError::io(&backup, err))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn write_json(path: &Path, value: Value) {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).unwrap();
        }
        fs::write(path, serde_json::to_vec_pretty(&value).unwrap()).unwrap();
    }

    fn write_plugin(root: &Path, harness: &str, name: &str) {
        let manifest_dir = match harness {
            "claude" => ".claude-plugin",
            "cursor" => ".cursor-plugin",
            "codex" => ".codex-plugin",
            _ => unreachable!(),
        };
        write_json(
            &root.join(manifest_dir).join("plugin.json"),
            serde_json::json!({"name": name, "version": "1.0.0"}),
        );
    }

    #[test]
    fn merges_three_marketplaces_for_one_logical_plugin() {
        let temp = tempfile::tempdir().unwrap();
        let root = temp.path().join("marketplace");
        let claude = root.join("providers/claude/plugin");
        let cursor = root.join("providers/cursor/plugin");
        let codex = root.join("providers/codex/plugin");
        write_plugin(&claude, "claude", "paddle");
        write_plugin(&cursor, "cursor", "paddle");
        write_plugin(&codex, "codex", "paddle");
        fs::create_dir_all(claude.join("skills/billing")).unwrap();
        fs::write(claude.join("skills/billing/SKILL.md"), "# Billing\n").unwrap();
        fs::create_dir_all(cursor.join("skills/billing")).unwrap();
        fs::write(cursor.join("skills/billing/SKILL.md"), "# Cursor billing\n").unwrap();
        fs::create_dir_all(codex.join("skills/billing")).unwrap();
        fs::write(codex.join("skills/billing/SKILL.md"), "# Codex billing\n").unwrap();
        fs::create_dir_all(cursor.join("rules")).unwrap();
        fs::write(cursor.join("rules/paddle.mdc"), "# Paddle\n").unwrap();
        fs::create_dir_all(codex.join("assets")).unwrap();
        fs::write(codex.join("assets/logo.svg"), "<svg/>\n").unwrap();

        write_json(
            &root.join(CLAUDE_MARKETPLACE),
            serde_json::json!({"plugins": [{"name": "paddle", "source": "./providers/claude/plugin"}]}),
        );
        write_json(
            &root.join(CURSOR_MARKETPLACE),
            serde_json::json!({"plugins": [{"name": "paddle", "source": "./providers/cursor/plugin"}]}),
        );
        write_json(
            &root.join(CODEX_MARKETPLACE),
            serde_json::json!({"plugins": [{"name": "paddle", "source": {"source": "local", "path": "./providers/codex/plugin"}}]}),
        );

        assert!(materialize_single_plugin_marketplace(&root).unwrap());
        assert!(root.join(".claude-plugin/plugin.json").is_file());
        assert!(root.join(".cursor-plugin/plugin.json").is_file());
        assert!(root.join(".codex-plugin/plugin.json").is_file());
        assert!(root.join("skills/billing/SKILL.md").is_file());
        assert_eq!(
            fs::read_to_string(root.join("skills/billing/SKILL.md")).unwrap(),
            "# Billing\n",
            "Claude is the deterministic portable baseline on cross-provider conflicts"
        );
        assert!(root.join("rules/paddle.mdc").is_file());
        assert!(root.join("assets/logo.svg").is_file());
        assert!(!root.join(CLAUDE_MARKETPLACE).exists());
        assert!(!materialize_single_plugin_marketplace(&root).unwrap());
    }

    #[test]
    fn supports_claude_plugin_root_metadata() {
        let temp = tempfile::tempdir().unwrap();
        let root = temp.path().join("marketplace");
        let plugin = root.join("plugins/paddle");
        write_plugin(&plugin, "claude", "paddle");
        write_json(
            &root.join(CLAUDE_MARKETPLACE),
            serde_json::json!({
                "metadata": {"pluginRoot": "./plugins"},
                "plugins": [{"name": "paddle", "source": "paddle"}]
            }),
        );

        assert!(materialize_single_plugin_marketplace(&root).unwrap());
        assert!(root.join(".claude-plugin/plugin.json").is_file());
    }

    #[test]
    fn rejects_multi_plugin_catalog_instead_of_flattening_it() {
        let temp = tempfile::tempdir().unwrap();
        let root = temp.path().join("marketplace");
        for name in ["one", "two"] {
            write_plugin(&root.join("plugins").join(name), "claude", name);
        }
        write_json(
            &root.join(CLAUDE_MARKETPLACE),
            serde_json::json!({"plugins": [
                {"name": "one", "source": "./plugins/one"},
                {"name": "two", "source": "./plugins/two"}
            ]}),
        );

        let error = materialize_single_plugin_marketplace(&root).unwrap_err();
        assert!(error.to_string().contains("multiple plugins (one, two)"));
        assert!(root.join(CLAUDE_MARKETPLACE).is_file());
    }

    #[test]
    fn rejects_source_path_traversal() {
        let temp = tempfile::tempdir().unwrap();
        let root = temp.path().join("marketplace");
        fs::create_dir_all(&root).unwrap();
        write_json(
            &root.join(CLAUDE_MARKETPLACE),
            serde_json::json!({"plugins": [{"name": "escape", "source": "./../escape"}]}),
        );

        let error = materialize_single_plugin_marketplace(&root).unwrap_err();
        assert!(error.to_string().contains("unsafe marketplace source"));
    }
}

File: internal/cache/marketplace.go
package cache

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

var pluginManifestPaths = []string{
	filepath.Join(".claude-plugin", "plugin.json"),
	filepath.Join(".cursor-plugin", "plugin.json"),
	filepath.Join(".codex-plugin", "plugin.json"),
}

type marketplaceSpec struct {
	relativePath string
	priority     uint8
}

var marketplaceSpecs = []marketplaceSpec{
	{relativePath: filepath.Join(".agents", "plugins", "marketplace.json"), priority: 0},
	{relativePath: filepath.Join(".cursor-plugin", "marketplace.json"), priority: 1},
	{relativePath: filepath.Join(".claude-plugin", "marketplace.json"), priority: 2},
}

type marketplaceDocument struct {
	Metadata struct {
		PluginRoot *string `json:"pluginRoot"`
	} `json:"metadata"`
	Plugins []marketplacePlugin `json:"plugins"`
}

type marketplacePlugin struct {
	Name   string          `json:"name"`
	Source json.RawMessage `json:"source"`
}

type localPluginSource struct {
	path     string
	priority uint8
}

func materializeSinglePluginMarketplace(root string) (bool, error) {
	if !hasMarketplaceManifest(root) {
		return false, nil
	}
	canonicalRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		return false, fmt.Errorf("resolve marketplace root %s: %w", root, err)
	}
	pluginNames := make(map[string]struct{})
	localSources := make(map[string]uint8)
	for _, spec := range marketplaceSpecs {
		manifestPath := filepath.Join(root, spec.relativePath)
		if !regularFile(manifestPath) {
			continue
		}
		var document marketplaceDocument
		if err := readJSON(manifestPath, &document); err != nil {
			return false, fmt.Errorf("invalid marketplace manifest %s: %w", manifestPath, err)
		}
		namesInDocument := make(map[string]struct{}, len(document.Plugins))
		for _, plugin := range document.Plugins {
			name := strings.TrimSpace(plugin.Name)
			if name == "" {
				return false, fmt.Errorf("marketplace manifest %s contains a plugin with an empty name", manifestPath)
			}
			if _, duplicate := namesInDocument[name]; duplicate {
				return false, fmt.Errorf("marketplace manifest %s contains duplicate plugin name %q", manifestPath, name)
			}
			namesInDocument[name] = struct{}{}
			pluginNames[name] = struct{}{}
			relativeSource, local, err := localSourcePath(plugin.Source, document.Metadata.PluginRoot, manifestPath, name)
			if err != nil {
				return false, err
			}
			if !local {
				continue
			}
			source, err := canonicalLocalSource(canonicalRoot, relativeSource, manifestPath, name)
			if err != nil {
				return false, err
			}
			if err := validatePluginSource(source, name, manifestPath); err != nil {
				return false, err
			}
			if existing, exists := localSources[source]; !exists || spec.priority > existing {
				localSources[source] = spec.priority
			}
		}
	}
	if len(pluginNames) == 0 {
		return false, fmt.Errorf("marketplace at %s contains no plugins", root)
	}
	if len(pluginNames) != 1 {
		names := mapKeys(pluginNames)
		return false, fmt.Errorf("marketplace at %s contains multiple plugins (%s); add a specific plugin source directory instead", root, strings.Join(names, ", "))
	}
	if len(localSources) == 0 {
		return false, fmt.Errorf("marketplace plugin %q at %s has no local source; add its Git or package source directly", mapKeys(pluginNames)[0], root)
	}
	sources := make([]localPluginSource, 0, len(localSources))
	for path, priority := range localSources {
		sources = append(sources, localPluginSource{path: path, priority: priority})
	}
	sort.Slice(sources, func(i, j int) bool {
		if sources[i].priority != sources[j].priority {
			return sources[i].priority < sources[j].priority
		}
		return sources[i].path < sources[j].path
	})
	parent := filepath.Dir(root)
	normalized, err := os.MkdirTemp(parent, ".agentpack-marketplace-normalized-")
	if err != nil {
		return false, fmt.Errorf("create normalized marketplace directory: %w", err)
	}
	installed := false
	defer func() {
		if !installed {
			_ = os.RemoveAll(normalized)
		}
	}()
	for _, source := range sources {
		if err := copyMergeTree(source.path, normalized); err != nil {
			return false, err
		}
	}
	if !hasAnyRelativeFile(normalized, pluginManifestPaths) {
		return false, fmt.Errorf("marketplace at %s did not resolve to a plugin manifest", root)
	}
	if err := replaceDirectory(root, normalized); err != nil {
		return false, err
	}
	installed = true
	return true, nil
}

func localSourcePath(raw json.RawMessage, pluginRoot *string, manifestPath, pluginName string) (string, bool, error) {
	var sourceText string
	trimmed := strings.TrimSpace(string(raw))
	switch {
	case strings.HasPrefix(trimmed, "\""):
		if err := json.Unmarshal(raw, &sourceText); err != nil {
			return "", false, fmt.Errorf("marketplace source for %q in %s must be a string or object", pluginName, manifestPath)
		}
	case strings.HasPrefix(trimmed, "{"):
		var object struct {
			Source string  `json:"source"`
			Path   *string `json:"path"`
		}
		if objectErr := json.Unmarshal(raw, &object); objectErr != nil {
			return "", false, fmt.Errorf("marketplace source for %q in %s must be a string or object", pluginName, manifestPath)
		}
		if object.Source != "local" {
			return "", false, nil
		}
		if object.Path == nil {
			return "", false, fmt.Errorf("local marketplace source for %q in %s has no string path", pluginName, manifestPath)
		}
		sourceText = *object.Path
	default:
		return "", false, fmt.Errorf("marketplace source for %q in %s must be a string or object", pluginName, manifestPath)
	}
	if pluginRoot == nil && !strings.HasPrefix(sourceText, "./") {
		return "", false, fmt.Errorf("local marketplace source %q for %q in %s must start with ./", sourceText, pluginName, manifestPath)
	}
	parts := make([]string, 0, 2)
	if pluginRoot != nil {
		clean, err := safeRelativePath(*pluginRoot, manifestPath, pluginName)
		if err != nil {
			return "", false, err
		}
		parts = append(parts, clean)
	}
	clean, err := safeRelativePath(sourceText, manifestPath, pluginName)
	if err != nil {
		return "", false, err
	}
	parts = append(parts, clean)
	relative := filepath.Join(parts...)
	if relative == "" || relative == "." {
		return "", false, fmt.Errorf("local marketplace source for %q in %s resolves to the marketplace root", pluginName, manifestPath)
	}
	return relative, true, nil
}

func safeRelativePath(value, manifestPath, pluginName string) (string, error) {
	normalized := strings.ReplaceAll(value, "\\", "/")
	if strings.HasPrefix(normalized, "/") || filepath.IsAbs(value) || filepath.VolumeName(value) != "" {
		return "", unsafeMarketplacePath(value, manifestPath, pluginName)
	}
	var parts []string
	for _, part := range strings.Split(normalized, "/") {
		switch part {
		case "", ".":
			continue
		case "..":
			return "", unsafeMarketplacePath(value, manifestPath, pluginName)
		default:
			parts = append(parts, part)
		}
	}
	return filepath.Join(parts...), nil
}

func unsafeMarketplacePath(value, manifestPath, pluginName string) error {
	return fmt.Errorf("unsafe marketplace source %q for %q in %s: paths must stay inside the marketplace root", value, pluginName, manifestPath)
}

func canonicalLocalSource(canonicalRoot, relativeSource, manifestPath, pluginName string) (string, error) {
	candidate := filepath.Join(canonicalRoot, relativeSource)
	canonical, err := filepath.EvalSymlinks(candidate)
	if err != nil {
		return "", fmt.Errorf("cannot resolve marketplace source for %q in %s at %s: %w", pluginName, manifestPath, candidate, err)
	}
	relative, err := filepath.Rel(canonicalRoot, canonical)
	if err != nil || relative == ".." || strings.HasPrefix(relative, ".."+string(filepath.Separator)) {
		return "", fmt.Errorf("marketplace source for %q in %s escapes the marketplace root: %s", pluginName, manifestPath, candidate)
	}
	info, err := os.Stat(canonical)
	if err != nil || !info.IsDir() {
		return "", fmt.Errorf("marketplace source for %q in %s is not a directory: %s", pluginName, manifestPath, candidate)
	}
	return canonical, nil
}

func validatePluginSource(source, expectedName, marketplace string) error {
	found := false
	for _, relative := range pluginManifestPaths {
		manifestPath := filepath.Join(source, relative)
		if !regularFile(manifestPath) {
			continue
		}
		found = true
		var manifest struct {
			Name *string `json:"name"`
		}
		if err := readJSON(manifestPath, &manifest); err != nil {
			return err
		}
		if manifest.Name == nil {
			return fmt.Errorf("plugin manifest %s referenced by %s has no string name", manifestPath, marketplace)
		}
		if *manifest.Name != expectedName {
			return fmt.Errorf("marketplace plugin name %q in %s does not match manifest name %q at %s", expectedName, marketplace, *manifest.Name, manifestPath)
		}
	}
	if !found {
		return fmt.Errorf("marketplace plugin %q in %s points to %s, which has no .claude-plugin, .cursor-plugin, or .codex-plugin manifest", expectedName, marketplace, source)
	}
	return nil
}

func replaceDirectory(current, replacement string) error {
	backupFile, err := os.CreateTemp(filepath.Dir(current), ".agentpack-marketplace-backup-")
	if err != nil {
		return fmt.Errorf("reserve marketplace backup: %w", err)
	}
	backup := backupFile.Name()
	if err := backupFile.Close(); err != nil {
		return fmt.Errorf("close marketplace backup reservation: %w", err)
	}
	if err := os.Remove(backup); err != nil {
		return fmt.Errorf("remove marketplace backup reservation: %w", err)
	}
	if err := os.Rename(current, backup); err != nil {
		return fmt.Errorf("move marketplace %s to backup: %w", current, err)
	}
	if err := os.Rename(replacement, current); err != nil {
		rollbackErr := os.Rename(backup, current)
		if rollbackErr != nil {
			return errors.Join(fmt.Errorf("install normalized marketplace: %w", err), fmt.Errorf("rollback marketplace: %w", rollbackErr))
		}
		return fmt.Errorf("install normalized marketplace: %w", err)
	}
	if err := os.RemoveAll(backup); err != nil {
		return fmt.Errorf("remove marketplace backup %s: %w", backup, err)
	}
	return nil
}

func readJSON(path string, destination any) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("read JSON %s: %w", path, err)
	}
	if err := json.Unmarshal(data, destination); err != nil {
		return fmt.Errorf("parse JSON %s: %w", path, err)
	}
	return nil
}

func hasAnyRelativeFile(root string, paths []string) bool {
	for _, relative := range paths {
		if regularFile(filepath.Join(root, relative)) {
			return true
		}
	}
	return false
}

func mapKeys(values map[string]struct{}) []string {
	keys := make([]string, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}
